File: code_cpp/self pased DSA/Graph/LEETCODE/3-digonal-traversal.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        
        int dirx[] = {0,1};
        int diry[] = {1,0};
        int n = mat.size();
        int m = mat[0].size();
        // apply BFS
        vector<vector<int>> vis(n, vector<int>(m, 0));
        
        vector<vector<int>> res;
        vector<int> ans;
        queue<vector<int>> q;
        q.push({0, 0});
        vis[0][0]=1;
        
        bool flag=false;
        while(!q.empty())
        {
            int size = q.size();
            vector<int> level;
            while(size-->0)
            {
                auto v = q.front();
                q.pop();
                
                int x = v[0], y = v[1];
                level.push_back(mat[x][y]);
                
                for(int i=0; i<2; i++)
                {
                    int nx = x + dirx[i];
                    int ny = y + diry[i];
                    // check if it is a valid 
                    if(nx>=0 && nx<n && ny>=0 && ny<m && !vis[nx][ny]){
                        q.push({nx,ny});
                        vis[nx][ny]=1;
                    }
                        
                }
            }
            if(flag==true){
                res.push_back(level);
                flag=false;
            }
            else{
                reverse(level.begin(), level.end());
                res.push_back(level);
                flag = true;
            }
        }
        
        for(auto v: res)
        {
            for(auto x: v){
                ans.push_back(x);
            }
        }
        return ans;
    }
};

File: code_java/self pased DSA/Graph/LEETCODE/3-digonal-traversal.java
/*
Problem: Diagonal Traversal of a Matrix
Given a matrix, return all elements of the matrix in diagonal order (zigzag level order traversal).

Approach:
- Use BFS to traverse the matrix level by level, alternating the order of each level (reverse every other level).
- Use a queue to process nodes and a visited matrix to avoid revisiting.
- Collect the result in a list.
*/

import java.util.*;

public class DiagonalTraversal {
    public List<Integer> findDiagonalOrder(int[][] mat) {
        int n = mat.length, m = mat[0].length;
        int[] dirx = {0, 1};
        int[] diry = {1, 0};
        boolean[][] vis = new boolean[n][m];
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> ans = new ArrayList<>();
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0, 0});
        vis[0][0] = true;
        boolean flag = false;
        while (!q.isEmpty()) {
            int size = q.size();
            List<Integer> level = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                int[] v = q.poll();
                int x = v[0], y = v[1];
                level.add(mat[x][y]);
                for (int d = 0; d < 2; d++) {
                    int nx = x + dirx[d], ny = y + diry[d];
                    if (nx >= 0 && nx < n && ny >= 0 && ny < m && !vis[nx][ny]) {
                        q.offer(new int[]{nx, ny});
                        vis[nx][ny] = true;
                    }
                }
            }
            if (flag) {
                res.add(level);
                flag = false;
            } else {
                Collections.reverse(level);
                res.add(level);
                flag = true;
            }
        }
        for (List<Integer> v : res) {
            ans.addAll(v);
        }
        return ans;
    }
} 